File: flutter_app/lib/models/conversation.dart
class Conversation {
  Conversation({
    required this.id,
    required this.otherUserId,
    required this.otherUserName,
    required this.unreadCount,
    this.lastMessage,
    this.lastMessageTime,
  });

  final int id;
  final int otherUserId;
  final String otherUserName;
  final int unreadCount;
  final String? lastMessage;
  final String? lastMessageTime;

  factory Conversation.fromJson(Map<String, dynamic> json) {
    return Conversation(
      id: _asInt(json['id']),
      otherUserId: _asInt(json['other_user_id']),
      otherUserName: (json['other_user_name'] ?? 'Unknown') as String,
      unreadCount: _asInt(json['unread_count']),
      lastMessage: json['last_message'] as String?,
      lastMessageTime: json['last_message_time'] as String?,
    );
  }
}

int _asInt(dynamic value) {
  if (value is int) {
    return value;
  }
  return int.tryParse('$value') ?? 0;
}

File: frontend/src/models/conversation.ts
import { asInt, asNullableString, asRecord, asString } from './parse';

export interface Conversation {
  id: number;
  otherUserId: number;
  otherUserName: string;
  unreadCount: number;
  lastMessage?: string;
  lastMessageTime?: string;
}

export function parseConversation(value: unknown): Conversation {
  const json = asRecord(value);
  return {
    id: asInt(json.id),
    otherUserId: asInt(json.other_user_id),
    otherUserName: asString(json.other_user_name, 'Unknown'),
    unreadCount: asInt(json.unread_count),
    lastMessage: asNullableString(json.last_message),
    lastMessageTime: asNullableString(json.last_message_time),
  };
}
